i18n(domains): validation and updated logic (#9099) (#9143)
withastro/astro: .changeset/tidy-carrots-jump.md

---
'@astrojs/vercel': minor
'@astrojs/node': minor
'astro': patch
---

> Add experimental support for i18n domain routing:

Reviewer: ```suggestion
Adds experimental support for a new i18n domain routing strategy (`"domain"`) that allows you to configure different domains for certain locales:
```

```js
// astro.config.mjs
import { defineConfig } from "astro/config"
export default defineConfig({
    experimental: {
        i18n: {
            defaultLocale: "en",
            locales: ["es", "en", "fr"],
            routingStrategy: "domain",
            domains: {
                fr: "https://fr.example.com",
                es: "https://example.es"
            }
        }
    }
})
```

> When this routing strategy is enabled and you run `astro build` command, the URLs emitted by `getAbsoluteLocaleUrl` and `getAbsoluteLocaleUrlList` will be different:
> 
> - `/fr/about` will become `https://fr.example.com/about`;
> - `/es/about` will become `https://example.es/about`;
> - `/en/about` will **stay** as is because we haven't configured any domain;

Reviewer: ```suggestion
With `routingStrategy: "domain"` configured, the URLs for your built site, including the URLs emitted by `getAbsoluteLocaleUrl()` and `getAbsoluteLocaleUrlList()`, will follow the pattern set for each locale individually.

For any `locales` not configured in `domains`, the fallback URLs created will follow `prefix-other-locales`, and create a `/[locale]/` path for all non-default languages.

For the above configuration:

- The file `/fr/about.astro` will create the URL `https://fr.example.com/about`;
- The file `/es/about.astro` will create the URL `https://example.es/about`;
- the file `/ja/about/astro` will create the URL `https://example.com/ja/about`.
- The file `/en/about.astro` will create the URL `https://example.com/about`.
```

So, I'm not sure whether this is true! But, I think adding `"ja"` to the list of locales and showing what happens would be very helpful. OR, specify that you are only allowed to not specify the defaultLocale, if that's the case. I think only the one example of "en" isn't enough to establish the pattern because it happens to be the default locale, so it could be a special case.

Author: Re 1.: good call out, I belive adding one more non-confiured locale in the example is perfect. 

Re 2: 
> I'm not sure whether you'd prefer to describe this as a "redirect" or "a file in this location produces this URL". Either is fine!

I prefer `"a file in this location produces this URL"` because it's closer to the actual behaviour. **We don't redirects**; we make assumptions, and developers must follow those assumptions if they want to have this feature.

```js
// astro.config.mjs
import { defineConfig } from "astro/config"
export default defineConfig({
  i18n: {
      defaultLocale: "en",
      locales: ["es", "en", "fr"],
      domains: {
          fr: "https://fr.example.com",
          es: "https://example.es"
      },
      routing: {
        prefixDefaultLocale: true,
        strategy: "domains"
      }
  },
  experimental: {
      i18nDomains: true
  },
  site: "https://example.com",
  output: "server"
})
```

With `routing.strategy` set to `"domains"`, the URLs emitted by `getAbsoluteLocaleUrl()` and `getAbsoluteLocaleUrlList()` will use the options set in `i18n.domains`:

```js
import { getAbsoluteLocaleUrl  } from "astro:i18n";

getAbsoluteLocaleUrl("en", "about"); // will return "https://example.com/en/about"
getAbsoluteLocaleUrl("fr", "about"); // will return "https://fr.example.com/about"
getAbsoluteLocaleUrl("es", "about"); // will return "https://example.es/about"
```

For the above configuration:

- The file `/fr/about.astro` will create the URL `https://fr.example.com/about`.
- The file `/es/about.astro` will create the URL `https://example.es/about`.

> - The file `/en/about.astro` will create the URL `https://example.com/en/about`.

Reviewer 2: I find this language a bit confusing since I don't know what "create the URL" means. 

I assume the intention is to describe that the server will handle requests to `https://fr.example.com/about` by rendering the `/fr/about.astro` file?

---
'astro': minor
---

Adds experimental support for a new i18n domain routing option (`"domains"`) that allows you to configure different domains for individual locales in entirely server-rendered projects.

To enable this in your project, first configure your `server`-rendered project's i18n routing with your preferences if you have not already done so. Then, set the `experimental.i18nDomains` flag to `true` and add `i18n.domains` to map any of your supported `locales` to custom URLs:

```js
//astro.config.mjs"
import { defineConfig } from "astro/config"
export default defineConfig({
  site: "https://example.com",

>   output: "server",

Reviewer: ```suggestion
  output: "server", // required, with no prerendered pages
```
Might as well do this here, too!

---

> '@astrojs/node': minor

Author: I reworked the changesets. This particular changeset is for `astro` only. I created a new changeset for `@astrojs/node` and `@astrojs/vercel` with a small sentence.
